core/src/middleware/mod.rs

```diff
@@ -296,14 +296,15 @@ impl<'a> BatchEntry<'a> {
 /// the underlying service requires that and otherwise one will get a compiler error that tries to
 /// explain that.
 pub trait RpcServiceT {
-	/// The error type.
-	type Error: std::fmt::Debug;
-
-	/// The response type
-	type Response: ToJson;
+	/// Response type for `RpcServiceT::call`.
+	type MethodResponse: ToJson;
+	/// Response type for `RpcServiceT::notification`.
+	type NotificationResponse: ToJson;
+	/// Response type for `RpcServiceT::batch`.
+	type BatchResponse: ToJson;
```

Comment:
iirc you said to me that `ToJson` is only needed in middlewares. If I got this right, perhaps only the middleware that needs it should have this `ToJson` constraint on things and the service itself shouldn't require it?

Reply:
good catch, fixed now